README.md

```diff
@@ -0,0 +1,443 @@
+Should these be different binaries (as they are now), or the same binary with different arguments?
+
+## Representing Phases with Types
+
+Olivier Gaça suggests that we could avoid warnings in comments to only call certain methods before/after startup by using distinct types to represent the component before and after startup.
+
+## ModuleParams Defaults
+
+Currently, the guide recommends that `ModuleParams` be optional and that its zero value be meaningful.
+But this may result in a lot of `if params.Xyz == "" { params.Xyz = "real default" }`, which could be annoying and ugly.
+If component users were encouraged to use `fx.Replace(foo.ModuleParams{ Xyz: "specific value" })` to specify parameters, then the component's module could use `fx.Supply(ModuleParams{ Xyz: "real default", ... }` to supply nonzero defaults.
+
+## Enabling / Disabling Component Bundles
+
+Consider configs like `logs_enabled` or `apm_config.enabled` -- when these are false, it means that several components that are already instantiated should not actually start.
+One options here is to thread a bunch of `Enable` methods through these components.
+Then an app would call
+```go
+fx.Invoke(func(config config.Component, agent logsagent.Component) {
+    if config.GetBool("logs_enabled") {
+        logsagent.Enable()
+    }
+})
+```
+
+and the logs agent's Enable method would call the Enable method on its internal components (schedulers, etc. -- anything that does something active).
+This approach is verbose -- writing `Enable` methods everywhere, adding an `enable bool` field, and then checking that field all over the place.
+It also complicates things like health monitoring, status, and ipcserver -- those all require registration during startup, _before_ it's known whether the component is enabled.
+So we'll need a way for all of those to handle disabled components.
+
+See [#2](https://github.com/djmitche/dd-agent-comp-experiments/pull/2) for an attempt at solving this (which fails because the Health registration comes after start).
+
+# TODO
+
+ * [DONE] component.yml?
+ * [DONE] actor model conventions
+ * [DONE] subscription conventions
+ * [DONE] guidelines for non-component stuff
+ * [DONE] health reporting
+ * [DONE] add Mocks to a component and try them out
+ * [DONE] selecting among multiple implementations of the same component (e.g., Tagger)
+ * [DONE] CLI / subcommands (`agent run`, etc.)
+ * [DONE] use fx.In for complex constructors (maybe with component concrete type?)
+ * [DONE] API
+ * [DONE] status output
+ * [DONE] update pkg/util/subscriptions to interact nicely with plugin registrations
+ * startup for bundles? the hidden Invoke() of a constructor that does nothing is weird.
+   * Doc how to handle enabling / disabling -- maybe this is per-bundle?
+ * more tests
+   * some kind of external test (`./test/`?)
+ * more components
+   * tagger
```

Comment:
Do we have a way to handle fallback from remote to local components? I think it's especially useful for hostname detection (as I think it basically blocks everything else), but maybe we can be confident enough in our code to think that if hostname detection fails in one process, it will fail everywhere.

Reply:
That's a much deeper question about remote components than I've considered so far.